pkg/image/trigger/buildconfigs/buildconfigs.go

```diff
@@ -0,0 +1,228 @@
+package buildconfigs
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+
+	"github.com/golang/glog"
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	triggerapi "github.com/openshift/origin/pkg/image/api/v1/trigger"
+	"github.com/openshift/origin/pkg/image/trigger"
+)
+
+// calculateBuildConfigTriggers transforms a build config into a set of image change triggers.
+func calculateBuildConfigTriggers(bc *buildapi.BuildConfig) []triggerapi.ObjectFieldTrigger {
+	var triggers []triggerapi.ObjectFieldTrigger
+	for _, t := range bc.Spec.Triggers {
+		if t.ImageChange == nil {
+			continue
+		}
+		var (
+			fieldPath string
+			from      *kapi.ObjectReference
+		)
+		if t.ImageChange.From != nil {
+			from = t.ImageChange.From
+			fieldPath = "spec.triggers"
+		} else {
+			from = buildutil.GetInputReference(bc.Spec.Strategy)
+			fieldPath = "spec.strategy.*.from"
+		}
+		if from == nil || from.Kind != "ImageStreamTag" || len(from.Name) == 0 {
+			continue
+		}
+
+		// add a trigger
+		triggers = append(triggers, triggerapi.ObjectFieldTrigger{
+			From: triggerapi.ObjectReference{
+				Name:       from.Name,
+				Namespace:  from.Namespace,
+				Kind:       from.Kind,
+				APIVersion: from.APIVersion,
+			},
+			FieldPath: fieldPath,
+		})
+	}
+	return triggers
+}
+
+func UpdateBuildConfigImages(bc *buildapi.BuildConfig, tagRetriever trigger.TagRetriever) (*buildapi.BuildConfig, error) {
```

Comment:
github was hiding the diff on me.

I don't care about the package, but i still disagree w/ code that's not used by the product being located in product source files.  it makes it easier to break behavior w/o realizing it because you change this code at the same time and inadvertently "fix" the test you're actually breaking.